lib/evaluate/type.h

```diff
@@ -35,168 +38,245 @@ namespace Fortran::evaluate {
+// category that could possibly be supported on any target.
+template<TypeCategory CATEGORY, int... KINDS>
+using CategoryTypesTuple = std::tuple<Type<CATEGORY, KINDS>...>;
+
+template<TypeCategory CATEGORY> struct CategoryTypesHelper;
+template<> struct CategoryTypesHelper<TypeCategory::Integer> {
+  using type = CategoryTypesTuple<TypeCategory::Integer, 1, 2, 4, 8, 16>;
+};
+template<> struct CategoryTypesHelper<TypeCategory::Real> {
+  using type = CategoryTypesTuple<TypeCategory::Real, 2, 4, 8, 10, 16>;
+};
+template<> struct CategoryTypesHelper<TypeCategory::Complex> {
+  using type = CategoryTypesTuple<TypeCategory::Complex, 2, 4, 8, 10, 16>;
+};
+template<> struct CategoryTypesHelper<TypeCategory::Character> {
+  using type = CategoryTypesTuple<TypeCategory::Character, 1>;  // TODO: 2 & 4
+};
+template<> struct CategoryTypesHelper<TypeCategory::Logical> {
+  using type = CategoryTypesTuple<TypeCategory::Logical, 1, 2, 4, 8>;
+};
+template<TypeCategory CATEGORY>
+using CategoryTypes = typename CategoryTypesHelper<CATEGORY>::type;
+
+using IntegerTypes = CategoryTypes<TypeCategory::Integer>;
+using RealTypes = CategoryTypes<TypeCategory::Real>;
+using ComplexTypes = CategoryTypes<TypeCategory::Complex>;
+using CharacterTypes = CategoryTypes<TypeCategory::Character>;
+using LogicalTypes = CategoryTypes<TypeCategory::Logical>;
+
+using FloatingTypes = common::CombineTuples<RealTypes, ComplexTypes>;
+using NumericTypes = common::CombineTuples<IntegerTypes, FloatingTypes>;
+using RelationalTypes = common::CombineTuples<NumericTypes, CharacterTypes>;
+using AllIntrinsicTypes = common::CombineTuples<RelationalTypes, LogicalTypes>;
+
+// When Scalar<T> is S, then TypeOf<S> is T.
+// TypeOf is implemented by scanning all supported types for a match
+// with Type<T>::Scalar.
+template<typename CONST> struct TypeOfHelper {
+  template<typename T> struct Predicate {
+    static constexpr bool value() {
+      return std::is_same_v<std::decay_t<CONST>,
+          std::decay_t<typename T::Scalar>>;
+    }
+  };
+  static constexpr int index{
+      common::SearchMembers<Predicate, AllIntrinsicTypes>};
+  using type = std::conditional_t<index >= 0,
+      std::tuple_element_t<index, AllIntrinsicTypes>, void>;
+};
+
+template<typename CONST> using TypeOf = typename TypeOfHelper<CONST>::type;
+
+// A variant union that can hold a scalar constant of any type chosen from
+// a set of types, which is passed in as a tuple of Type<> specializations.
+template<typename TYPES> struct SomeScalar {
+  using Types = TYPES;
+  CLASS_BOILERPLATE(SomeScalar)
+
+  template<typename A> SomeScalar(const A &x) : u{x} {}
   template<typename A>
-  GenericScalar(std::enable_if_t<!std::is_reference_v<A>, A> &&x)
+  SomeScalar(std::enable_if_t<!std::is_reference_v<A>, A> &&x)
     : u{std::move(x)} {}
 
-  std::optional<std::int64_t> ToInt64() const;
-  std::optional<std::string> ToString() const;
+  auto ToInt64() const {
```

Comment:
I think these functions would be more readable with explicit return types.

Reply:
OK.